File: 2017/day04/day04.py
day_num = "04"
day_title = "High-Entropy Passphrases"


def read_input(filename):
    with open(filename) as f:
        content = [x.strip('\n') for x in f.readlines()]
    return content


def isValid(passphrase: str) -> bool:
    words = passphrase.split()
    s = set()
    for w in words:
        if w in s:
            return False
        s.add(w)
    return True


def isValidNoAnagrams(passphrase: str) -> bool:
    words = passphrase.split()
    for i in range(len(words)-1):
        for j in range(i+1, len(words)):
            if isAnagram(words[i], words[j]):
                return False
    return True


def isAnagram(a: str, b: str) -> bool:
    return sorted(list(a)) == sorted(list(b))


def part1():
    passphrases = read_input('input.txt')
    count = 0
    anagram_count = 0
    for phrase in passphrases:
        if isValid(phrase):
            count += 1
        if isValidNoAnagrams(phrase):
            anagram_count += 1

    print(f"Part 1: {count}")
    print(f"Part 2: {anagram_count}")


def main():
    print("Day", day_num, ":", day_title)
    part1()


if __name__ == '__main__':
    main()

File: 2023/day04/day04.go
package main

import (
	"fmt"
	"strings"

	"github.com/icemanblues/advent-of-code/pkg/util"
)

const (
	dayNum   = "04"
	dayTitle = "Scratchcards"
)

type IntSet map[int]struct{}

func (is IntSet) intersection(os IntSet) int {
	count := 0
	for k := range os {
		if _, ok := is[k]; ok {
			count++
		}
	}
	return count
}

type Scratchcard struct {
	ID            int
	Winning, Hand IntSet
}

func (sc Scratchcard) matches() int {
	return sc.Winning.intersection(sc.Hand)
}

func (sc Scratchcard) score() int {
	matched := sc.matches()
	if matched > 0 {
		return util.IntPow(2, matched-1)
	}
	return 0
}

func NewIntSet(s []string) IntSet {
	is := make(IntSet)
	for _, r := range s {
		i := util.MustAtoi(string(r))
		is[i] = struct{}{}
	}
	return is
}

func parseScratchcards(filename string) []Scratchcard {
	input, _ := util.ReadInput(filename)
	sc := make([]Scratchcard, 0, len(input))
	for _, line := range input {
		idWinHand := strings.Split(line, ": ")
		id := util.MustAtoi(strings.Fields(idWinHand[0])[1])

		winHands := strings.Split(idWinHand[1], " | ")
		win, hands := NewIntSet(strings.Fields(winHands[0])), NewIntSet(strings.Fields(winHands[1]))
		sc = append(sc, Scratchcard{ID: id, Winning: win, Hand: hands})
	}
	return sc
}

func part1() {
	scratchcards := parseScratchcards("input.txt")
	sum := 0
	for _, sc := range scratchcards {
		sum += sc.score()
	}
	fmt.Printf("Part 1: %v\n", sum)
}

func part2() {
	scratchcards := parseScratchcards("input.txt")
	cardCounts := make([]int, len(scratchcards), len(scratchcards))
	for i := range cardCounts {
		cardCounts[i] = 1
	}
	for i, sc := range scratchcards {
		mult := cardCounts[i]
		for j := 0; j < sc.matches(); j++ {
			if i+j+1 >= len(cardCounts) {
				break
			}
			cardCounts[i+j+1] += mult
		}
	}
	sum := 0
	for _, v := range cardCounts {
		sum += v
	}
	fmt.Printf("Part 2: %v\n", sum)
}

func main() {
	fmt.Printf("Day %v: %v\n", dayNum, dayTitle)
	part1()
	part2()
}
